c/uca/sifter/unisifeggs.cpp

```diff
@@ -0,0 +1,433 @@
+#include <algorithm>
+#include <array>
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <ranges>
+#include <string>
+#include <string_view>
+#include <vector>
+
+class CheckFailure {
+ public:
+  ~CheckFailure() {
+    std::terminate();
+  }
+  void operator&(std::ostream&) {}
+};
+
+#define CHECK(condition) \
+  (condition)            \
+      ? (void)0          \
+      : CheckFailure() & std::cerr << ("Check failed: " #condition "\n")
+
+class CodePointRange {
+ public:
+  constexpr CodePointRange(char32_t front, char32_t back)
+      : front_(front), back_(back) {
+    if (empty()) {
+      front_ = 1;
+      back_ = 0;
+    }
```

Comment:
Suggestion:
- rename "front" to "start"
- rename "back" to "end"
- internally store "start_" and "limit_" where limit_ = end + 1
- empty() is start_ == limit_ or maybe start_ >= limit_

This from experience working with the API and innards of UnicodeSet. Exclusive-end "limit" values tend to be easier to work with than inclusive-end "end" values.

Reply:
See my other comments (though would be fine with storing begin_ and end_ rather than front_ and back_).
`begin` and `end` in particular are needed and need to have the right semantics for range-based for loops to work (but more generally, having the right names is required for code that works with containers to work with this).

Reply:
Grumble C++ STL grumble...

Please
- name the fields start_ and end_
- add a limit() function that returns end_ + 1, and use that where it makes sense

Reply:
Note that I changed the type to be an iterator, rather than a char32_t; so using iterator terminology probably makes sense (the choice of type also meants that if you construct from a range [c, d[, you need to construct it from iterators rather than code points, which makes it clearer that the second bound is excluded).

Reply:
Then again it means I am mixing iterators across range objects in some operations, which feels wrong. I will think about that some more another day.

Reply:
Can you live with the names `first`, `past_the_end` (setting aside the 🐪 vs. 🐫 vs. 🐍 discussion for a moment), and `last` when we have non-API names corresponding to the standard `*begin=front`, `end`, and `back=*--end`?

That is consistent with wording in the standard, specifically, 
https://eel.is/c++draft/container.requirements.general#container.reqmts-28,
https://eel.is/c++draft/container.requirements.general#container.reqmts-31, and
https://eel.is/c++draft/container.requirements#sequence.reqmts-115,
and I think it should be clearer to the reader unfamiliar with standard library naming conventions.

Reply:
first and last would be ok.

I could live with past_the_end because it's clear, but it's clunky. We have been using "limit" in ICU for nearly 30 years...

Reply:
(rant)
Gotta love this one:
> https://eel.is/c++draft/container.requirements#container.reqmts-31

“**b.end()** ... Returns: An iterator which is the **past-the-end** value for the container.”

So end() does not return an iterator *at* the end, it returns one *past the end*. Then why not call it for what it does??
https://www.google.com/search?q=dunce+corner&tbm=isch
(/rant)

Reply:
> (rant)

Yeah, it’s silly. (_end_, from _past-the-end_ by apheresis, like _bus_ from _omnibus_? 🙃)